File: sqg_run.py
from sqgturb import SQG, SQGState
import jax.numpy as jnp
import numpy as np

# run SQG turbulence simulation, save data to netcdf file

# model parameters.

#N = 512 # number of grid points in each direction (waves=N/2)
#dt = 90 # time step in seconds
#diff_efold = 1800. # time scale for hyperdiffusion at smallest resolved scale

#N = 192
#dt = 300
#diff_efold = 86400./8.
 
#N = 128
#dt = 600
#diff_efold = 86400./3.

#N = 96
#dt = 900
#diff_efold = 86400./2.

N = 64
dt = 1200    
diff_efold = 86400. 

norder = 8 # order of hyperdiffusion

# Ekman damping coefficient r=dek*N**2/f, dek = ekman depth = sqrt(2.*Av/f))
# Av (turb viscosity) = 2.5 gives dek = sqrt(5/f) = 223
# for ocean Av is 1-5, land 5-50 (Lin and Pierrehumbert, 1988)
# corresponding to ekman depth of 141-316 m over ocean.
# spindown time of a barotropic vortex is tau = H/(f*dek), 10 days for
# H=10km, f=0.0001, dek=100m.
dek = 0 # applied only at surface if symmetric=False
nsq = 1.e-4; f=1.e-4; g = 9.8; theta0 = 300
H = 10.e3 # lid height
r = dek*nsq/f
U = 20 # jet speed
Lr = np.sqrt(nsq)*H/f # Rossby radius
L = 20.*Lr
# thermal relaxation time scale
tdiab = 10.*86400 # in seconds
# parameter used to scale PV to temperature units.
scalefact = f*theta0/g

# create random noise
rs = np.random.RandomState(42) # fixed seed
pv = rs.normal(0,100.,size=(2,N,N)).astype(np.float32)
# add isolated blob on lid
nexp = 20
x = np.arange(0,2.*np.pi,2.*np.pi/N); y = np.arange(0.,2.*np.pi,2.*np.pi/N)
x,y = np.meshgrid(x,y)
x = x.astype(np.float32); y = y.astype(np.float32)
pv[1] = pv[1]+2000.*(np.sin(x/2)**(2*nexp)*np.sin(y)**nexp)
# remove area mean from each level.
for k in range(2):
    pv[k] = pv[k] - pv[k].mean()

# initialize sqg model instance
model = SQG(N,nsq=nsq,f=f,U=U,H=H,r=r,tdiab=tdiab,dt=dt,
            diff_order=norder,diff_efold=diff_efold,
            precision='single')

#  initialize figure.
outputinterval = 6.*3600. # interval between frames in seconds
tmin = 100.*86400. # time to start saving data (in days)
tmax = 430.*86400. # time to stop (in days)
nsteps = int(tmax/outputinterval) # number of time steps to animate
# set number of timesteps to integrate for each call to model.advance
ntimesteps = int(outputinterval/model.dt)
savedata = 'sqgu%s_N%s_6hrly_24hdiff_jax.nc' % (U,N) # save data plotted in a netcdf file.

if savedata is not None:
    from netCDF4 import Dataset
    nc = Dataset(savedata, mode='w', format='NETCDF4_CLASSIC')
    nc.r = model.r[0]
    nc.f = model.f
    nc.U = model.U
    nc.L = model.L
    nc.H = model.H
    nc.g = g; nc.theta0 = theta0
    nc.nsq = model.nsq
    nc.tdiab = model.tdiab
    nc.dt = model.dt
    nc.diff_efold = model.diff_efold
    nc.diff_order = model.diff_order
    x = nc.createDimension('x',N)
    y = nc.createDimension('y',N)
    z = nc.createDimension('z',2)
    t = nc.createDimension('t',None)
    pvvar =\
    nc.createVariable('pv',np.float32,('t','z','y','x'),zlib=True)
    pvvar.units = 'K'
    # pv scaled by g/(f*theta0) so du/dz = d(pv)/dy
    xvar = nc.createVariable('x',np.float32,('x',))
    xvar.units = 'meters'
    yvar = nc.createVariable('y',np.float32,('y',))
    yvar.units = 'meters'
    zvar = nc.createVariable('z',np.float32,('z',))
    zvar.units = 'meters'
    tvar = nc.createVariable('t',np.float32,('t',))
    tvar.units = 'seconds'
    xvar[:] = np.arange(0,model.L,model.L/N)
    yvar[:] = np.arange(0,model.L,model.L/N)
    zvar[0] = 0; zvar[1] = model.H

t = 0.0
state = SQGState(
        pvspec=jnp.fft.rfft2(jnp.array(pv)),
        t=t)
nout = 0
while t < tmax:
    state, pv = model.advance(state,timesteps=ntimesteps)
    t = state.t
    hr = t/3600.
    print(hr,scalefact*pv.min(),scalefact*pv.max())
    if savedata is not None and t >= tmin:
        print('saving data at t = %g hours' % hr)
        pvvar[nout,:,:,:] = np.asarray(pv)
        tvar[nout] = t
        nc.sync()
        if t >= tmax: nc.close()
        nout = nout + 1

File: sqg_run.cpp
/**
 * sqg_run.cpp
 *
 * Example driver for the C++ SQG model, mirroring sqg_run.py.
 * Runs the model and prints hourly statistics to stdout.
 *
 * Build with CMake:
 *   mkdir build && cd build && cmake .. && make -j
 *   ./sqg_run
 */
#include "sqg.h"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <random>
#include <vector>

int main()
{
    // ------------------------------------------------------------------
    // Model parameters (match sqg_run.py defaults)
    // ------------------------------------------------------------------
    const int   N          = 96;
    const float dt         = 900.0f;          // seconds
    const float diff_efold = 86400.0f / 2.0f; // half a day
    const int   norder     = 8;
    const bool  dealias    = true;

    const float dek    = 0.0f;                // Ekman depth (zero → no Ekman)
    const float nsq    = 1.0e-4f;
    const float f      = 1.0e-4f;
    const float g      = 9.8f;
    const float theta0 = 300.0f;
    const float H      = 10.0e3f;
    const float r      = dek * nsq / f;
    const float U      = 20.0f;              // jet speed (m/s)
    const float Lr     = std::sqrt(nsq) * H / f;
    const float L      = 20.0f * Lr;
    const float tdiab  = 10.0f * 86400.0f;
    const bool  sym    = true;

    const float scalefact = f * theta0 / g;

    const double tmax           = 300.0 * 86400.0;
    const double output_interval = 6.0  * 3600.0;

    // ------------------------------------------------------------------
    // Initialise PV: random noise + isolated blob on the lid
    // ------------------------------------------------------------------
    std::vector<float> pv(2 * N * N, 0.0f);

    std::mt19937                    rng(42);
    std::normal_distribution<float> noise(0.0f, 100.0f);

    for (int i = 0; i < 2 * N * N; i++)
        pv[i] = noise(rng);

    // Add isolated blob on lid (level 1)
    const int nexp = 20;
    for (int iy = 0; iy < N; iy++) {
        float y = 2.0f * static_cast<float>(M_PI) * iy / N;
        for (int ix = 0; ix < N; ix++) {
            float x   = 2.0f * static_cast<float>(M_PI) * ix / N;
            float blob = 2000.0f
                       * std::pow(std::sin(x / 2.0f), 2 * nexp)
                       * std::pow(std::sin(y),          nexp);
            pv[1 * N * N + iy * N + ix] += blob;
        }
    }

    // Remove area mean from each level
    for (int lev = 0; lev < 2; lev++) {
        float sum = 0.0f;
        for (int k = 0; k < N * N; k++) sum += pv[lev * N * N + k];
        float mean = sum / static_cast<float>(N * N);
        for (int k = 0; k < N * N; k++) pv[lev * N * N + k] -= mean;
    }

    // ------------------------------------------------------------------
    // Construct model
    // ------------------------------------------------------------------
    SQG model(pv.data(), N,
              f, nsq, L, H, U, r, tdiab,
              norder, diff_efold,
              theta0, g,
              sym,
              dt,
              dealias,
              /*threads=*/1,
              /*tstart=*/0.0f);

    // Set number of sub-steps per advance() call
    model.timesteps = static_cast<int>(std::round(output_interval / dt));

    // ------------------------------------------------------------------
    // Time loop
    // ------------------------------------------------------------------
    std::vector<float> pv_out(2 * N * N);

    std::printf("%-12s  %-14s  %-14s  %-14s\n",
                "time(hr)", "max_speed", "pv_min(K)", "pv_max(K)");

    while (model.t < tmax) {
        model.advance(pv_out.data());

        double hr = model.t / 3600.0;

        // Max wind speed (level 1 = lid)
        float spd_max = 0.0f;
        for (int k = 0; k < N * N; k++) {
            float uu = model.u[N * N + k];
            float vv = model.v[N * N + k];
            float spd = std::sqrt(uu * uu + vv * vv);
            if (spd > spd_max) spd_max = spd;
        }

        // PV extrema on lid (level 1), converted to temperature units (K)
        float pv_min =  1.0e30f;
        float pv_max = -1.0e30f;
        for (int k = 0; k < N * N; k++) {
            float val = scalefact * pv_out[N * N + k];
            if (val < pv_min) pv_min = val;
            if (val > pv_max) pv_max = val;
        }

        std::printf("%-12.2f  %-14.4f  %-14.4f  %-14.4f\n",
                    hr, spd_max, pv_min, pv_max);
        std::fflush(stdout);
    }

    return 0;
}
